src/charm_state.py

```diff
@@ -1071,6 +1072,25 @@ def _check_immutable_config_change(
         except KeyError as exc:
             logger.info("Key %s not found, this will be updated to current config.", exc.args[0])
 
+    @classmethod
+    def _log_prev_state(cls, prev_state_dict: dict) -> None:
+        """Log the previous state of the charm.
+
+        Replace sensitive information before logging.
+
+        Args:
+            prev_state_dict: The previous state of the charm as a dict.
+        """
+        if logger.isEnabledFor(logging.DEBUG):
+            prev_state_for_logging = prev_state_dict.copy()
+            charm_config = prev_state_for_logging.get("charm_config")
+            if charm_config and "token" in charm_config:
+                charm_config = charm_config.copy()
+                charm_config["token"] = "*****"  # nosec
+            prev_state_for_logging["charm_config"] = charm_config
+
+            logger.debug("Previous charm state: %s", prev_state_for_logging)
+
```

Comment:
This might not hide all token during integration tests, as we supply multiple tokens during integration test and only pick one to use. There are also `alt-token` in integration tests.
For the `tox` process these tokens would still be visible.

Reply:
I think the charm state only contains the `token` config. You are referring to the `tox` command line which may be in the runner's workload, which should be hidden using the `hide_cmd=True` setting for the `ps aux` command.